add FLIP for cadence mutability restrictions (#703)
onflow/flow: flips/20211129-cadence-mutability-restrictions.md

# Title of FLIP

| Status        | (Proposed)       |
:-------------- |:---------------------------------------------------- |

> | **FLIP #**    | [NNN](https://github.com/onflow/flow/pull/NNN) (update when you have PR #)|

Author: ```suggestion
| **FLIP #**    | [703](https://github.com/onflow/flow/pull/703)      |
```

| **Author(s)** | Daniel Sainati (daniel.sainati@dapperlabs.com)       |
| **Sponsor**   | Daniel Sainati (daniel.sainati@dapperlabs.com)       |

> | **Updated**   | 2021-11-29                                           |

Author: ```suggestion
| **Updated**   | 2021-12-08                                           |
```

However, simply writing to a field directly is not the only way in which one can modify a value. Consider the following example:

```
pub struct Foo {
    pub let x : [Int]

    init() {
        self.x = [3];
    }
}

pub fun bar() {
    let foo = Foo()
    foo.x = [0] // writes to x, not allowed
    foo.x[0] = 0; // does not write to x, also not allowed
}
```

Cadence also restricts the scopes in which an array or dictionary field can be modified (or "mutated"). Examples of 
mutating operations include an indexed assignment, as in the above example, as well as calls to the `append` or `remove`,
methods of arrays, or the `insert` or `remove` methods of dictionaries. These operations can only be performed on a field
in the current and inner scopes, the same contexts in which the field could be written to if it were a `var`. So the following 
would typecheck:

```
pub struct Foo {
    pub let x : [Int]

    init() {
        self.x = [3];
    }

    pub fun addToX(i: Int) {
        self.x.append(i)
    }
}
```

while the following would not:

```
pub struct Foo {
    pub let y : [Int]

    init() {
        self.y = [3];
    }
}

pub fun addToY(foo: Foo, i: Int) {
    foo.y.append(i)

> }

Author: Yeah I describe this below, but the builtin array/dictionary functions are given a bit tracking whether they mutate the receiver or not.

```

This prevents external code from mutating the values of fields it can read from your contract. Consumers
of your contract may read the values in a `pub let` or `pub var` field, but cannot change them in any way. 

If you wish to allow other code to update or modify a field in your contract, you may expose a method 
to do so, like in the example above with `addToX`, or you may use the `pub(set)` access mode, which 
allows any code to mutate or write to the field it applies to. 

### Implementation Details 

This change adds a new error, the `ExternalMutationError`, which is raised when a field 
is mutated outside of the context in which it was defined. The error message will also
suggest that the user instead use a setter or modifier method for that field.

Specifically, the error is emitted whenever a user attempts to perform an 

> index assignment on a member that is not either declared with the `pub(set)` 

Reviewer: Does the proposal specifically only handle arrays and dictionaries, and the indexing assignment, or does it also address field assignments on composite values?
For example:

```rust
pub struct Foo {
  pub var name: String
  init() {
    self.name = "initial"
  }
}

pub struct Bar {
  pub let foo: Foo
  init() {
    self.foo = Foo()
  }
}

let bar = Bar()
bar.foo.name = "new"  // is this also rejected?
```

Reviewer 2: I think this is the case : 
`pub struct Foo {
  pub(set) var name: String
  init() {
    self.name = "initial"
  }
}

pub struct Bar {
  pub let foo: Foo
  init() {
    self.foo = Foo()
  }
}

let bar = Bar()
bar.foo.name = "new"  // is this also rejected?`

Reviewer: @dsainati1 Sorry, I wrote the code in the comment field and missed the essential part, like @bluesign corrected, `Foo.name` should be `pub(set)`. 

The point I'm trying to make here: Just like the proposal is currently proposing to improve working with containers such as arrays and dictionaries, which are mutable types, composites can also be mutable.

Author: Yea this is an interesting point, and one where I think there is room for more discussion on the design. It is certainly possible that we should restrict writes to composites like this: i.e. if any of the members in an `x.y.z = e` chain are declared with something either than `pub(set) var`, we might not allow them to be mutated. However, given that `pub(set) var`s must be explicitly declared by users, I don't believe they as likely to be accidentally mutated. 

In my opinion, I think we can use the same behavior for mutations we currently use for writes, wherein a `pub(set)` field can still be written to even if it is a member of a composite that is stored in a `let` field. However, I am curious what arguments there are for a different approach.

Reviewer 3: Would this apply to calling functions on composite fields also? I don't think it would, but maybe something worth considering

Reviewer: I like the proposal, but it seems like it only addresses part of the general problem, i.e. only focuses on arrays and dictionaries, but composite values are also containers. 

Then again, maybe it is good to keep the scope of this proposal narrow, and we can extend the new behaviour to composites in a follow-up proposal.

Reviewer 4: I'm really on the fence here, and the example in this document really makes it a tough call.

```
    foo.y.append(1) // invalid, field `y` is let and cannot be mutated
```

In this case, I'm not assigning to `y`, I'm calling a method on `y`. If `y` were a user-defined type with a method called `append()`, this would NOT be an error. To make it an error just because `y` is an language-defined collection type is pretty inconsistent.

OTOH- Containers are the place where this kind of "gotcha" is most likely to be a problem, so maybe it's worth introducing an inconsistency temporarily until we can find a better solution?

I don't feel confident enough to push this discussion in either direction, to be honest.

Author: Looking at this and @turbolent's comments below, the consensus seems to be that this FLIP does not address 100% of the footgun problem, but does hit the most common cases. If we want to handle the parts of this "gotcha" case, as I see it we have a couple options:

1) We can update the FLIP to handle the nested-composite mutation case. There are a couple ways I can see to to do this: we add a `mutating` keyword (or similar) to method definitions, and prevent such methods from being called on fields declared with a `let`. We would also have the option of enforcing that `mutating` methods are the only way to modify internal state (like Swift does), or we could choose to allow the modifier to be used at the developer's discretion, which is easier but also less accurate. 

2) We merge the FLIP now and attempt to address the nested-composite case later. The difficulty here is that any later solution to the nested-composite case would have to be a non-breaking change. Adding a `mutating` keyword as described above, but only restricting where `mutating` methods are used rather than also requiring certain methods to be marked as `mutating` is one way to address this without a breaking change, but it certainly isn't a perfect solution.

3) We merge the FLIP as-is, and just live with the inconsistency/incompleteness. 

My personal preference is for option 2, as long as we can come up with a non-breaking solution that we can apply later that is also a satisfying way to solve the problem. Interested to see if anybody has any thoughts on non-breaking fixes for this.

Reviewer 3: I'm fine with just not allowing anything on the array and dictionary fields, since like Dete said, that is the place where the vast majority of people make their mistakes. The inconsistency doesn't bother me much.

access mode, or is defined in the current enclosing scope. This check is the
same one performed for writing to fields, with the difference that mutation 
is allowed on both `let` and `var` fields, while only the latter can be written to.

Additionally, array and dictionary methods now track an additional bit of information
indicating whether they mutate their receiver. Mutating methods may not be called on 
members that are not declared with the `pub(set)` access mode, or defined in the current

> enclosing scope. 

Reviewer: Would we want to expose this to Cadence programs? Swift has the `mutating` function modifier for this purpose.

Just curious, I don't think that is necessary for this proposal, it could be added through a future proposal.

Author: We could; unsure what the use case is for this, but it's possible.

The array methods that are considered mutating are `append`, `appendAll`, `remove`,
`insert`, `removeFirst` and `removeLast`. The dictionary methods that are considered

> mutating are `insert` and `remove`. 

Reviewer: Will only these methods, or also the operators performing these operations be considered mutating? For example, `let xs: {"a": 1}; xs["a"] = 2`. I think they should and probably are implicit here, but maybe mention it explicitly

Author: Yes indexed assignment is also considered mutating.

themselves in the foot by exporting a `pub let` field from their contract and having it be externally mutated 
if `pub let` fields cannot exist on contracts in the first place. However, this has a number of downsides. The
obvious one is that all data that one might wish to expose to be read from a contract would need an explicit getter
method. The second, less obvious but also more concerning downside, is that it would be necessary to 
ban `pub let` fields on structs and resources as well. Consider the case where a user exports a getter method
to read a struct or resource used by their contract. Any `pub let` fields on this struct or resource 
would also be mutable, and thus are subject to the same risks they would be if they were exported
directly from the contract. 

Consider:

```
pub contract C {
    pub struct Foo {
        pub let arr : [Int]
        init() {
            self.arr = [3]
        }
    }
    
    priv let foo : Foo

    init() {
        self.foo = Foo()
    }

    pub fun getFoo(): Foo {
        return self.foo
    }
}

pub fun main() {
    let a = C.getFoo()

>     a.arr.append(0) // a.arr is now [3, 0]

Reviewer: Is this what is proposed here? Currently, this should not be expected, as the field's value is copied

Author: When I run this code as a contract/transaction pair on the playground, the output is `[3,0]`. Is that not expected?

```
pub contract C {
    pub struct Foo {
        pub let arr : [Int]
        init() {
            self.arr = [3]
        }
    }
    
    priv let foo : Foo
    init() {
        self.foo = Foo()
    }
    pub fun getFoo(): Foo {
        return self.foo
    }
}
```
and 
```
import C from 0x01

transaction {

  prepare(acct: AuthAccount) {}

  execute {
    let a = C.getFoo()
    a.arr.append(0) // a.arr is now [3, 0]
    log(a.arr)
  }
}
```

Reviewer: Ah, my bad, I misread the example as the array being returned from the getter and being modified!

}
```

Another approach would be to have fields use a different type depending on the scope in which they are 
viewed. Within a struct, a `pub let` array field would have the type `[Int]`, permitting all array 
operations on it, including mutating ones. Outside the struct, the same field might have the type 
`[Int]{ReadonlyArray<Int>}`, in effect a restricted type that only permits the non-mutating 
operations to be performed on that array. This, however, is overly complex, and introduces
a new type to accomplish the same purpose as what this FLIP proposes. 

### Performance Implications

* This is a change only to the Cadence type checker, so it will not have any 

> performance implications for Flow or the Cadence runtime. 

Reviewer: As type checking is currently run as part of execution, it implicitly has an effect on total performance of the system

consumers of their code to be able to modify. While in the past, a field declared
with `pub let` could be mutated, now users will need to provide modifier or setter
functions on their contracts or other structures if they wish to allow this behavior. 
This will be communicated via changes to the Cadence documentation. 

### Examples

Some examples of code that produces a type error as a result of this restriction:

```
pub resource Foo {
    pub let x : {Int: Int}

    init() {
        self.x = {0:3};
    }
}

pub fun bar() {
    let foo <- create Foo()
    foo.x[0] = 3 // cannot mutate `x`, field was defined in `Foo`
    destroy foo
}
```

```
pub struct Bar {
    pub let foo: Foo
    init() {
        self.foo = Foo()
    }
}

pub struct Foo {
    pub let x : [Int]

    init() {
        self.x = [3]
    }
}

pub fun bar() {
    let bar = Bar()
    bar.foo.x[0] = 3 // cannot mutate `x`, field was defined in `Foo`
}
```

```
pub contract Foo {
    pub let x : S
    
    pub struct S {
        pub let y : [Int]
        init() {
            self.y = [3]
        }
    }

    init() {
        self.x = S()
    }
}

pub fun bar() {

>     Foo.x.y.remove(at: 0) // cannot mutate `y`, field was defined in `S`

Author: here `Foo` is a contract, which does not need to be instantiated like a struct or resource would be

	}
}

// in some external code importing Foo
pub fun bar() {
	Foo.x.append(1)
}
```

Currently Cadence does not warn against this, or prevent a developer from writing this code, even
though, depending on what is stored in `x`, this could be unsafe. 

## User Benefit

This change makes it less likely for a user to expose a value from a contract or other
publicly accessible structure that is not intended to be mutable, but can be modified 
by consumers of their contract regardless. This is important for a future stable, 
permissionless version of Cadence, where users do not need to have their smart
contracts reviewed in order to deploy them to Flow. By removing one possible way that 
users could write unsafe smart contracts, we make it easier for everyone to trust the
correctness of contracts deployed to Flow. 

## Design Proposal

### Explanation 

Cadence controls where and to what extent variables can be read and written using a combination of
access modifiers and declaration kinds, as described [here](https://docs.onflow.org/cadence/language/access-control/).
Of note is that the `let` kind does not allow fields to be written to in any scope, whereas `var` allows them

> to be written in the "Current and Inner" scopes; that is, the scope in which the field was declared, and any scopes

Reviewer: "written to" sounds like mutation is not allowed, but in fact it is, and this is exactly what this proposal is trying to introduce.

Maybe reword to something like "reassigned", etc.

contained within that scope. 

However, simply writing to a field directly is not the only way in which one can modify a value. Consider the following example:

```
pub struct Foo {
    pub let x: [Int]

    init() {
        self.x = [3]
    }
}

pub fun bar() {
    let foo = Foo()

>     foo.x = [0] // writes to x, not allowed
>     foo.x[0] = 0 // does not write to x, currently allowed

Reviewer: ```suggestion
    foo.x = [0] // reassignment to `x` is not allowed, as variable is `let`
    foo.x[0] = 0 // mutates the array, but does not reassign to `x`, currently allowed
```

}
```

This FLIP would change Cadence to also restrict the scopes in which an array or dictionary field can be modified (or "mutated"). 
Examples of mutating operations include an indexed assignment, as in the above example, as well as calls to the `append` or `remove`,
methods of arrays, or the `insert` or `remove` methods of dictionaries. These operations will only be able to be performed on a field
in the current and inner scopes, the same contexts in which the field could be written to if it were a `var`. So the following 
would typecheck:

> ```
> pub struct Foo {
>     pub let x : [Int]
> 
>     init() {
>         self.x = [3];
>     }

Reviewer: ```suggestion
```cadence
pub struct Foo {
    pub let x: [Int]

    init() {
        self.x = [3]
    }
```

    pub fun addToX(i: Int) {

>         self.x.append(i)

Reviewer: ```suggestion
        self.x.append(i)  // mutation of field `x` in same scope is valid, even though field is `let`
```

    }
}
```

while the following would not:

> ```
> pub struct Foo {
>     pub let y : [Int]
> 
>     init() {
>         self.y = [3];

Reviewer: ```suggestion
```cadence
pub struct Foo {
    pub let y: [Int]

    init() {
        self.y = [3]
```

    }
}

> pub fun addToY(foo: Foo, i: Int) {
>     foo.y.append(i)
> }

Reviewer: Maybe demonstrate using a "script function" and add an explanation comment:

```suggestion
pub fun main() {
    let foo = Foo()
    foo.y.append(1) // invalid, field `y` is let and cannot be mutated
}
```

allows any code to mutate or write to the field it applies to. 

### Implementation Details 

This change adds a new error, the `ExternalMutationError`, which is raised when a field 
is mutated outside of the context in which it was defined. The error message will also
suggest that the user instead use a setter or modifier method for that field.

Specifically, the error is reported whenever a user attempts to perform an 
index assignment on a member that is not either declared with the `pub(set)` 
access mode, or is defined in the current enclosing scope. This check is the
same one performed for writing to fields, with the difference that mutation 
is allowed on both `let` and `var` fields, while only the latter can be written to.

Additionally, array and dictionary methods now track an additional bit of information
indicating whether they mutate their receiver. Mutating methods may not be called on 
members that are not declared with the `pub(set)` access mode, or defined in the current
enclosing scope. 

The array methods that are considered mutating are `append`, `appendAll`, `remove`,
`insert`, `removeFirst` and `removeLast`. The dictionary methods that are considered

> mutating are `insert` and `remove`. Additionally, indexed assignment (`x[y] = e`) will
> be considered mutating for both arrays and dictionaries.  

Reviewer: ```suggestion
mutating are `insert` and `remove`. Additionally, indexed assignment (e.g. `x[y] = e`) will
be considered mutating for both arrays and dictionaries.  
```

Maybe mention that this will apply to all assignment operators, e.g. also the swap operator (`<->`)

The limitations on mutation are designed to closely mirror the limitations on writes, 
so that they can be easily explained to and understood by the user in terms of 
language principles with which they are already familiar. Similarly, the suggested
workaround of adding a setter or modifier method to the composite type is designed 
to be immediately recognizable to any developer familiar with object-oriented
design principles. 

### Drawbacks

This has the potential to break a number of existing contracts by restricting 
code that was previously legal. This change would require a migration path for 
these contracts in order for developers to update their contracts to satisfy
Cadence's new restrictions.

This also removes a small amount of expressivity from the language. Previously, 
the `pub let` declaration was a way to create a field that could be read and
mutated in all contexts, but not written, while `pub(set) var` created
a field that could be read, written and mutated in all contexts. After this change,

> it would not be possible to describe a field that can be read and mutated but not set, 

Reviewer 3: Are there cases where this is needed that a setter function would be inadequate?

Author: IMO, no, there are no specific cases where you could not accomplish the same thing with a setter function, but I felt it was worth at least mentioning, since it is a concept that can no longer be expressed.

a complex initialization for a dictionary or array value may benefit from the ability to iteratively mutate its contents,
up to the point at which its value is finalized, and can thus be exported. As such we would like to still allow users
to perform mutations in a limited context.

A third approach would be to ban contract-level public field, with the idea that users cannot accidentally shoot 
themselves in the foot by exporting a `pub let` field from their contract and having it be externally mutated 
if `pub let` fields cannot exist on contracts in the first place. However, this has a number of downsides. The
obvious one is that all data that one might wish to expose to be read from a contract would need an explicit getter
method. The second, less obvious but also more concerning downside, is that it would be necessary to 
ban `pub let` fields on structs and resources as well. Consider the case where a user exports a getter method
to read a struct or resource used by their contract. Any `pub let` fields on this struct or resource 
would also be mutable, and thus are subject to the same risks they would be if they were exported
directly from the contract. 

Consider:

> ```
> pub contract C {
>     pub struct Foo {
>         pub let arr : [Int]
>         init() {
>             self.arr = [3]
>         }
>     }
>     
>     priv let foo : Foo

Reviewer: ```suggestion
```cadence
pub contract C {
    pub struct Foo {
        pub let arr: [Int]
        init() {
            self.arr = [3]
        }
    }
    
    priv let foo: Foo
```

operations on it, including mutating ones. Outside the struct, the same field might have the type 
`[Int]{ReadonlyArray<Int>}`, in effect a restricted type that only permits the non-mutating 
operations to be performed on that array. This, however, is overly complex, and introduces
a new type to accomplish the same purpose as what this FLIP proposes. 

### Performance Implications

* This is a change to the Cadence type checker, so by adding more checks and errors
for it to report, it may have a performance impact. However, this should not add
signficantly more work to typechecking of programs that succeed on typechecks. 

### Best Practices

* This changes the way that users will need to expose fields that they wish for
consumers of their code to be able to modify. While in the past, a field declared
with `pub let` could be mutated, now users will need to provide modifier or setter
functions on their contracts or other structures if they wish to allow this behavior. 
This will be communicated via changes to the Cadence documentation. 

### Examples

Some examples of code that produces a type error as a result of this restriction:

> ```
> pub resource Foo {
>     pub let x : {Int: Int}
> 
>     init() {
>         self.x = {0:3};

Reviewer: ```suggestion
```cadence
pub resource Foo {
    pub let x: {Int: Int}

    init() {
        self.x = {0: 3}
```

    }
}

pub fun bar() {
    let foo <- create Foo()
    foo.x[0] = 3 // cannot mutate `x`, field was defined in `Foo`
    destroy foo
}
```

```
pub struct Bar {
    pub let foo: Foo
    init() {
        self.foo = Foo()
    }
}

pub struct Foo {

>     pub let x : [Int]

Reviewer: ```suggestion
    pub let x: [Int]
```

    init() {
        self.x = [3]
    }
}

pub fun bar() {
    let bar = Bar()
    bar.foo.x[0] = 3 // cannot mutate `x`, field was defined in `Foo`
}
```

> ```
> pub contract Foo {
>     pub let x : S
>     
>     pub struct S {
>         pub let y : [Int]
>         init() {
>             self.y = [3]

Reviewer: ```suggestion
```cadence
pub contract Foo {
    pub let x: S
    
    pub struct S {
        pub let y: [Int]
        init() {
            self.y = [3]
```

        }
    }

    init() {
        self.x = S()
    }
}

pub fun bar() {
    Foo.x.y.remove(at: 0) // cannot mutate `y`, field was defined in `S`
}
```

> ```
> pub contract Foo {
>     pub let x : S
>     
>     pub struct S {
>         pub let y : [Int]

Reviewer: ```suggestion

```cadence
pub contract Foo {
    pub let x: S
    
    pub struct S {
        pub let y: [Int]
```

        init() {
            self.y = [3]
        }
    }

    init() {
        self.x = S()
        self.x.y.append(2) // cannot mutate `y`, field was defined in `S`
    }
}	
```

while the following are allowed:

> ```
> pub struct Foo {
>     pub let x: {Int: Int}
> 
>     init() {
>         self.x = {3:3};

Reviewer: ```suggestion
```cadence
pub struct Foo {
    pub let x: {Int: Int}

    init() {
        self.x = {3: 3}
```

    }

    pub fun bar() {
        let foo = Foo()
        foo.x[0] = 3 // ok, mutation occurs inside defining struct Foo
    }
}
```

> ```
> pub struct Foo {
>     pub(set) var x: {Int: Int}
> 
>     init() {
>         self.x = {3:3};

Reviewer: ```suggestion

```cadence
pub struct Foo {
    pub(set) var x: {Int: Int}

    init() {
        self.x = {3: 3}
```

}

pub fun bar() {
    let foo = Foo()
    foo.x.insert(key: 0, 3) // ok, pub(set) access modifier allows mutation
}
```

### Compatibility

This change is not backwards compatible with existing Cadence smart contracts, but
is being made in service of a future version of Cadence wherein no backwards incompatible
changes will be required going forwards. It should not affect any other parts of Flow, 

> however. 

Reviewer 3: ```suggestion
```
However probably isn't necessary?

### User Impact

This will change the way users write their smart contracts. Given that this is preventing
users from writing an unsafe pattern, we do not expect this to invalidate a particularly
large amount of existing code. However, any code that is invalidated by this change can be
refactored to expose a setter/modifier method for affected fields. 

## Related Issues

Do we wish to handle aliased references? Consider the following example (courtesy of @SupunS):

> ```

Reviewer: ```suggestion
```cadence
```

pub fun main() {
  let foo <- create Foo()

  log(foo.immutable)

  foo.mutable.content = "updated" // mutating the 'immutable' via 'mutable'

  log(foo.immutable)

  destroy foo
}

pub resource Foo {
  pub(set) var mutable: &Bar

  pub let immutable: &Bar

  init() {
    self.mutable = &Bar() as &Bar

    self.immutable = self.mutable.   // immutable holds a direct/indirect reference to a mutable
  }
}

pub struct Bar {
  pub(set) var content: String
  init() {
    self.content = "original"
  }
}
```

This is one potential way to get around mutability restrictions on `pub let` fields, by aliasing such

> a field with a `pub(set) var`. Is preventing this kind of issue within the scope of this change? How

Reviewer 3: This is analogous to a setter function to me, so it seems fine to allow this

Cadence Mutability Restrictions

> | Status        | (Proposed)       |

Author: ```suggestion
| Status        | (Approved)       |
```
